### DISCUSSION TITLE
is reading ListView.SelectedItem thread-safe?

### DISCUSSION THREAD

**Discussion by deivid1** [tui-cs/Terminal.Gui#1444] [Q&A]
Could it happen that I read the SelectedItem value of a ListView and at the same time it's been modified so it creates some kind of issue? What if I want to change SelectedItem not manually but in the code?

**Comment by tznind** [comment]
What collection are you using underneath the `ListView`? `SelectedItem` is an integer so the index may not be valid if you are making lots of changes to your collection in multiple threads.  But if that's the case you might also run into issues during redraw etc.

If you want to be threadsafe then I would stick to assigning a new list and using `MainLoop.Invoke`.  That should eliminate any kind of out of bounds risk.  Since you only ever change the source in the `Invoke` bit it will be guaranteed not to coincide with any other events (e.g. keypress, mouse click etc).

Here is an example below:

```csharp
Application.Init ();
var w = new Window () { Width = Dim.Fill (), Height = Dim.Fill () };

ListView lv = new ListView () { Width = Dim.Fill (), Height = Dim.Fill () };
w.Add (lv);

// this is a Thread that will be accumulating new items
Task.Run (() => {

	List<int> newList = new List<int> ();
	Random r = new Random ();
	DateTime lastRefresh = DateTime.Now;

	while(true) {
		//50 milliseconds
		Task.Delay (50).Wait();
		newList.Add (r.Next ());

		// Every 5 seconds
		if(DateTime.Now - lastRefresh > TimeSpan.FromSeconds(5)) {
			
			lastRefresh = DateTime.Now ;

			// take the current state of the list and write it to the ListView but do it on the main UI thread
			Application.MainLoop.Invoke (() => {

				// maintain the selection / scroll index when updating
				var i = lv.SelectedItem;
				var top = lv.TopItem;

				// create a copy of the list to display in the UI
				var copy = newList.ToList ();
				lv.SetSource (copy);

				lv.SelectedItem = i;
				lv.TopItem = top;
			});
		}
	}
});
```

**Reply by deivid1** [reply]
Hi, thanks for your answer. My question is even more simple: ListView doesn't change its content at all. So If I move the selection with the keyboard the selected item change. My question is that if in a thread I read lv.SelectedItem at the same time that Terminal Gui is changing that SelectedItem variable, could it create a issue? like when you read and write a variable in two different tasks without using any lock.

**Reply by tznind** [reply]
Ah ok.   Then no, I can't see any way that that could cause you an issue.  There is nothing special about the `SelectedItem` 'get' accessor that could cause any problems.

I think in general you only need to worry about this kind of thing when you are counting e.g. reading then incrementing (`x++` vs `Interlocked.Increment(ref x)`).  It sounds like your Thread is just reading the value not even changing it so either way you will get an accurate value.  It will be before or after the keyboard keypress depending on when the user presses the button and how often the Thread polls. 

Now if your thread was constantly updating SelectedItem then you could probably see the potential for wierdness like you could get two lines in a row rendered as selected theoretically.  The way to avoid that kind of thing would be to use `MainLoop.AddIdle` for your second 'Thread' (but again that would only apply if you were changing the value not reading it).
